File: com/exnw/browedit/renderutils/Shader.java
package com.exnw.browedit.renderutils;

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.nio.ByteBuffer;
import java.nio.IntBuffer;
import java.util.HashMap;
import java.util.Map;

import javax.media.opengl.GL4;

import com.exnw.browedit.grflib.GrfLib;

public class Shader
{
	int shaderprogram;
	String vsrc;
	String fsrc;
	String vertexShaderFileName;
	String fragmentShaderFileName;
	
	Map<String, Uniform> uniforms = new HashMap<String, Uniform>();
	
	
	public Shader(String vertexShader, String fragmentShader, GL4 gl)
	{
		this.vertexShaderFileName = vertexShader;
		this.fragmentShaderFileName = fragmentShader;
		try
		{
			BufferedReader brv = new BufferedReader(new InputStreamReader(GrfLib.openFile(vertexShader)));
			vsrc = "";
			String line;
			while ((line=brv.readLine()) != null) {
			  vsrc += line + "\n";
			}
	
			BufferedReader brf = new BufferedReader(new InputStreamReader(GrfLib.openFile(fragmentShader)));
			fsrc = "";
			while ((line=brf.readLine()) != null) {
			  fsrc += line + "\n";
			}
		}
		catch(IOException e)
		{
			e.printStackTrace();
		}
	}
	
	void compile(GL4 gl)
	{
		int v = gl.glCreateShader(GL4.GL_VERTEX_SHADER);
		int f = gl.glCreateShader(GL4.GL_FRAGMENT_SHADER);
		gl.glShaderSource(v, 1, new String[]{vsrc}, IntBuffer.wrap(new int[] { vsrc.length() }));
		gl.glCompileShader(v);
		printShaderInfoLog(v,vertexShaderFileName, gl);
		gl.glShaderSource(f, 1, new String[]{fsrc}, IntBuffer.wrap(new int[] { fsrc.length() }));
		gl.glCompileShader(f);
		printShaderInfoLog(f, fragmentShaderFileName, gl);

		shaderprogram = gl.glCreateProgram();
		gl.glAttachShader(shaderprogram, v);
		gl.glAttachShader(shaderprogram, f);
	}
	
	void link(GL4 gl)
	{
		gl.glLinkProgram(shaderprogram);
		gl.glValidateProgram(shaderprogram);
	}
	
	
	public void printShaderInfoLog(int shader, String filename, GL4 gl)
	{
		IntBuffer b = IntBuffer.allocate(2);
		gl.glGetShaderiv(shader, GL4.GL_INFO_LOG_LENGTH,b);
		
		if (b.get(0) > 1)
		{
			byte[] info = new byte[b.get(0)];
			gl.glGetShaderInfoLog(shader, b.get(0), b, ByteBuffer.wrap(info));
			System.err.println("Error in " + filename + "\n" + new String(info));
		}
	}
	
	public Uniform getUniform(String name, GL4 gl)
	{
		if(!uniforms.containsKey(name))
		{
			int location = getUniformLocation(name, gl);
			if(location >= 0)
				uniforms.put(name, new Uniform(this, location, name));
			else
			{
				System.err.println("Could not find uniform " + name + " in shader => " + location);
			//	System.exit(-1);
			}
		}
		if(uniforms.containsKey(name))
			return uniforms.get(name);
		return null;
	}
	
	
	int getUniformLocation(String location, GL4 gl)
	{
		return gl.glGetUniformLocation(shaderprogram, location);
	}
	
	void bindAttributeLocation(String name, int position, GL4 gl)
	{
		gl.glBindAttribLocation(shaderprogram, position, name);
	}
	
	
	private String[] attributes = { "a_position", "a_normal", "a_color", "a_texcoord", "a_texcoord2" };
	
	public int getAttribLocation(String name)
	{
		for(int i = 0; i < attributes.length; i++)
			if(attributes[i].equals(name))
				return i;
		System.err.println("THIS SHOULDN'T HAPPEN");
		return -1;
	}
	

	public void use(GL4 gl)
	{
		if(shaderprogram == 0)
		{
			compile(gl);
			for(int i = 0; i < attributes.length; i++)
				bindAttributeLocation(attributes[i], i, gl);
			link(gl);
		}
		gl.glUseProgram(shaderprogram);		
	}
}

File: BroLib/Brolib/Shader.cpp
#include "Shader.h"
#include <Brolib/fs.h>
#include <BroLib/Log.h>

#include <glm/gtc/type_ptr.hpp>


Shader::Shader(std::string vertexShader, std::string fragmentShader)
{
	programId = glCreateProgram();
	shaders.push_back(new SubShader(vertexShader, SubShader::Vertex));
	shaders.push_back(new SubShader(fragmentShader, SubShader::Fragment));

	for(std::list<SubShader*>::iterator it = shaders.begin(); it != shaders.end(); it++)
		(*it)->attach(programId);
}


Shader::~Shader(void)
{
	glDeleteProgram(programId);
}

void Shader::use()
{
	glUseProgram(programId);
}

void Shader::link()
{
	glLinkProgram(programId);
}




Shader::SubShader::SubShader(std::string filename, ShaderType shaderType)
{
	shaderId = glCreateShader(shaderType == Vertex ? GL_VERTEX_SHADER : GL_FRAGMENT_SHADER);
	std::string data = File::getData(filename);
	const char* cdata = data.c_str();
	glShaderSource(shaderId, 1, &cdata,NULL);
	glCompileShader(shaderId);
	int status;
	glGetShaderiv(shaderId, GL_COMPILE_STATUS, &status);
	if(status == GL_FALSE)
	{
		int length, charsWritten;
		glGetShaderiv(shaderId, GL_INFO_LOG_LENGTH, &length);
		char* infolog = new char[length+1];
		memset(infolog, 0, length+1);
		glGetShaderInfoLog(shaderId, length, &charsWritten, infolog);
		logger<<"Error compiling shader "<<filename<<":\n"<<infolog<<Log::newline;
		delete[] infolog;
	}
}


Shader::SubShader::~SubShader()
{
	glDeleteShader(shaderId);
}

void Shader::SubShader::attach(GLuint programId)
{
	glAttachShader(programId, shaderId);
}


void Shader::bindAttributeLocation(std::string name, int position)
{
	glBindAttribLocation(programId, position, name.c_str());
}

void Shader::bindFragLocation( std::string name, int position )
{
	if(glBindFragDataLocation)
		glBindFragDataLocation(programId, position, name.c_str());
}




GLuint Shader::getUniformLocation( std::string name )
{
	if(uniformLocations.find(name) == uniformLocations.end())
		uniformLocations[name] = glGetUniformLocation(programId, name.c_str());
	return uniformLocations[name];
}




void Shader::setUniform( std::string name, const glm::mat4& value )
{
	int location = getUniformLocation(name);
	glUniformMatrix4fv(location, 1, 0, glm::value_ptr(value));
}
